launch_ros/launch_ros/ros_adapters.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2018-2020 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module containing ROS specific adapters and their initialization."""
+
+import os
+import threading
+from typing import List
+from typing import Optional
+
+import launch
+import launch.events
+
+import rclpy
+from rclpy.executors import SingleThreadedExecutor
+
+
+class ROSAdapter:
+    """Wraps rclpy API to ease ROS node usage in `launch_ros` actions."""
+
+    def __init__(
+        self,
+        *,
+        argv: Optional[List[str]] = None,
+        context: Optional[rclpy.Context] = None,
+        autostart: bool = True
+    ):
+        """
+        Construct adapter.
+
+        :param: argv List of global arguments for rclpy context initialization.
+        :param: context Provide a context other than the default rclpy context
+          to pass down to rclpy.init.
+          The context is expected to have already been initialized by the caller
+          using rclpy.init().
```

Comment:
This is something I wanted to discuss. We were already inconsistent about initializing / shutting down contexts. I think it'd be better to own these contexts and their lifetime, and always initialize them and shut them down unless explicitly stated otherwise. Thoughts?

Reply:
A context of its own is now instantiated for the adapter in 577d77bfc5fa2101616ce7856b6d55745370692b. Its lifetime is now entirely handled by the adapter.